mount/mount_linux.go

```diff
@@ -292,18 +297,71 @@ func (mounter *SafeFormatAndMount) checkAndRepairXfsFilesystem(source string) er
 			return nil
 		} else {
 			klog.Warningf("Filesystem corruption was detected for %s, running xfs_repair to repair", source)
-			out, err := mounter.Exec.Command("xfs_repair", args...).CombinedOutput()
-			if err != nil {
-				return NewMountError(HasFilesystemErrors, "'xfs_repair' found errors on device %s but could not correct them: %s\n", source, out)
-			} else {
-				klog.Infof("Device %s has errors which were corrected by xfs_repair.", source)
-				return nil
+			var uncorrectedErr error
+
+			// If xfs_repair fails to repair the file system successfully, try giving the same xfs_repair
+			// command twice more; xfs_repair may be able to make more repairs on successive runs.
+			for i := 0; i < 3; i++ {
+				out, err := mounter.Exec.Command("xfs_repair", args...).CombinedOutput()
+				if err != nil {
+					if e, isExitError := err.(utilexec.ExitError); isExitError {
+						switch e.ExitStatus() {
+						case xfsRepairErrorsUncorrected:
+							uncorrectedErr = NewMountError(HasFilesystemErrors, "'xfs_repair' found errors on device %s but could not correct them: %s\n", source, string(out))
+						// If the exit status is 2, do not retry, replay the dirty logs instead.
+						case xfsRepairErrorsDirtyLogs:
+							return mounter.replayXfsDirtyLogs(source)
+						default:
+							return NewMountError(HasFilesystemErrors, "'xfs_repair' found errors on device %s but could not correct them: %s\n", source, string(out))
+						}
+					} else {
+						return NewMountError(HasFilesystemErrors, "failed to run 'xfs_repair' on device %s: %v\n", source, err)
+					}
+				} else {
+					klog.Infof("Device %s has errors which were corrected by xfs_repair.", source)
+					return nil
+				}
+			}
+			if uncorrectedErr != nil {
+				return uncorrectedErr
 			}
 		}
 	}
 	return nil
 }
 
+// replayXfsDirtyLogs tries to replay dirty logs by by mounting and
+// immediately unmounting the filesystem on the same class of machine
+// that crashed.
+func (mounter *SafeFormatAndMount) replayXfsDirtyLogs(source string) error {
+	klog.V(4).Infof("Attempting to replay the dirty logs on device %s", source)
+	msg := fmt.Sprintf("failed to replay dirty logs on device %s", source)
+	target, err := ioutil.TempDir("", "")
+	if err != nil {
+		return NewMountError(HasFilesystemErrors, "%s: %v\n", msg, err)
+	}
+	defer os.RemoveAll(target)
+
+	klog.V(4).Infof("Attempting to mount disk %s at %s", source, target)
+	if err := mounter.Interface.Mount(source, target, "", []string{"defaults"}); err != nil {
```

Comment:
Why not use `target` field given in the mount command as mount location rather than tempdir? Could a kubelet/app crash leave the volume mounted in tempdir and never get cleaned up?

Reply:
I think maybe the target is not existing or not available for mounting for some potential reasons. tempdir is more safer to temporarily mount and unmount, and we will make sure it is deleted in the end.
But you are right, it is a risk if the caller crashes between mount and unmount.

Reply:
Think again, it is the caller of formatAndMount's responsibility to make sure the existence and availablity of the target, so I think we can use the `target` field here.